common/thorhelper/thorread.cpp

```diff
@@ -0,0 +1,825 @@
+        if (actualFilter.numFilterFields())
+        {
+            const RtlRecord &actual = actualDiskMeta->queryRecordAccessor(true);
+            unsigned numOffsets = actual.getNumVarFields() + 1;
+            size_t * variableOffsets = (size_t *)alloca(numOffsets * sizeof(size_t));
+            RtlRow row(actual, nullptr, numOffsets, variableOffsets);
+            row.setRow(buffer, 0);  // Use lazy offset calculation
+            return actualFilter.matches(row);
+        }
+        else
+            return true;
+    }
+
+    size32_t getFixedDiskRecordSize();
+
+protected:
+    MemoryBuffer tempOutputBuffer;
+    MemoryBufferBuilder bufferBuilder;
+    ISourceRowPrefetcher * actualRowPrefetcher = nullptr;
+    Owned<ITranslator> translators;
+    const IDynamicTransform * translator = nullptr;
+    const IKeyTranslator * keyedTranslator = nullptr;
+    RowFilter actualFilter;               // This refers to the actual disk layout
+    bool eogPending = false;
+    bool needToTranslate;
+};
+
+
+BinaryDiskRowReader::BinaryDiskRowReader(unsigned _expectedCrc, IOutputMetaData & _expected, unsigned _projectedCrc, IOutputMetaData & _projected, unsigned _actualCrc, IOutputMetaData & _actual, IPropertyTree * options)
+: LocalDiskRowReader(_expectedCrc, _expected, _projectedCrc, _projected, _actualCrc, _actual, options), bufferBuilder(tempOutputBuffer, 0)
+{
+    RecordTranslationMode translationMode = (RecordTranslationMode)options->getPropInt("translationMode", (int)RecordTranslationMode::All);
+    translators.setown(getTranslators("BinaryDiskRowReader", expectedCrc, &_expected, actualCrc, &_actual, projectedCrc, &_projected, translationMode));
+    actualRowPrefetcher = actualDiskMeta->createDiskPrefetcher();
+    if (translators)
+    {
+        translator = &translators->queryTranslator();
+        keyedTranslator = translators->queryKeyedTranslator();
+    }
+    needToTranslate = (translator && translator->needsTranslate());
+}
+
+
+void BinaryDiskRowReader::clearInput()
+{
+    DiskRowReader::clearInput();
+    eogPending = false;
+}
+
+bool BinaryDiskRowReader::matches(const char * format, bool streamRemote, unsigned _expectedCrc, IOutputMetaData & _expected, unsigned _projectedCrc, IOutputMetaData & _projected, unsigned _actualCrc, IOutputMetaData & _actual, IPropertyTree * options)
+{
+    if (!strieq(format, "thor") && !strieq(format, "flat"))
+        return false;
+    return LocalDiskRowReader::matches(format, streamRemote, _expectedCrc, _expected, _projectedCrc, _projected, _actualCrc, _actual, options);
+}
+
+bool BinaryDiskRowReader::setInputFile(IFile * inputFile, const char * _logicalFilename, unsigned _partNumber, offset_t _baseOffset, IPropertyTree * meta, const FieldFilterArray & expectedFilter)
+{
+    if (!LocalDiskRowReader::setInputFile(inputFile, _logicalFilename, _partNumber, _baseOffset, meta, expectedFilter))
+        return false;
+
+    actualFilter.clear();
+    actualFilter.appendFilters(expectedFilter);
```

Comment:
just to check, actual=expect filter here, because rows will already have been translated by time actualFilter used in this context, right?